File: app/src/main/java/ua/com/todd/smsforwading/IncomingSmsBroadcastReceiver.java
package ua.com.todd.smsforwading;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.SmsMessage;

import java.sql.SQLException;

import ua.com.todd.smsforwading.data.HelperFactory;

public class IncomingSmsBroadcastReceiver extends BroadcastReceiver {
    @Override
    public void onReceive(Context context, Intent intent) {
        Bundle bundle = intent.getExtras();
        if (bundle != null) {
            Object[] pdus = (Object[]) bundle.get("pdus");
            final SmsMessage[] messages = new SmsMessage[pdus.length];
            for (int i = 0; i < pdus.length; i++) {
                messages[i] = SmsMessage.createFromPdu((byte[]) pdus[i]);
            }
            if (messages.length > 0) {
                StringBuilder sb = new StringBuilder();
                for (SmsMessage s : messages) {
                    sb.append(s.getMessageBody());
                    sb.append("\n\n");
                }
                Sms sms = new Sms(sb.toString());
                try {
                    HelperFactory.getHelper().getSmsDAO().create(sms);
                    Intent i = new Intent(context, MailSenderService.class);
                    context.startService(i);
                } catch (SQLException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}

File: app/src/main/kotlin/ua/com/todd/smsforwading/IncomingSmsBroadcastReceiver.kt
package ua.com.todd.smsforwading

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.telephony.SmsMessage

import java.sql.SQLException

import ua.com.todd.smsforwading.data.HelperFactory

public class IncomingSmsBroadcastReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val bundle = intent.getExtras()
        if (bundle != null) {
            val pdus = bundle.get("pdus") as Array<Any>
            val messages = arrayOfNulls<SmsMessage>(pdus.size)
            for (i in pdus.indices) {
                messages[i] = SmsMessage.createFromPdu(pdus[i] as ByteArray)
            }
            if (messages.size() > 0) {
                for (s in messages) {
                    val sms = Sms(s?.getMessageBody())
                    try {
                        HelperFactory.getHelper().getSmsDAO()?.create(sms)
                        context.startService(Intent(context, javaClass<MailSenderService>()))
                    } catch (e: SQLException) {
                        e.printStackTrace()
                    }
                }
            }
        }
    }
}
